test/common/grpc/buffered_async_client_test.cc

```diff
@@ -23,117 +27,145 @@ namespace Envoy {
 namespace Grpc {
 namespace {
 
+constexpr uint32_t kPendingBufferSizeLimit = 1 << 16;
+
 class BufferedAsyncClientTest : public testing::Test {
 public:
   BufferedAsyncClientTest()
-      : method_descriptor_(helloworld::Greeter::descriptor()->FindMethodByName("SayHello")) {
+      : api_(Api::createApiForTest()), dispatcher_(api_->allocateDispatcher("test_thread")),
+        method_descriptor_(helloworld::Greeter::descriptor()->FindMethodByName("SayHello")) {
     config_.mutable_envoy_grpc()->set_cluster_name("test_cluster");
 
     cm_.initializeThreadLocalClusters({"test_cluster"});
     ON_CALL(cm_.thread_local_cluster_, httpAsyncClient()).WillByDefault(ReturnRef(http_client_));
   }
 
+  void SetUp() override {
+    EXPECT_CALL(http_client_, start(_, _)).WillOnce(Return(&http_stream_));
+    EXPECT_CALL(http_stream_, sendHeaders(_, _));
+    EXPECT_CALL(http_stream_, reset());
+
+    raw_client_ = std::make_shared<AsyncClientImpl>(cm_, config_, test_time_.timeSystem());
+    client_ = std::make_unique<AsyncClient<helloworld::HelloRequest, helloworld::HelloReply>>(
+        raw_client_);
+  }
+
+  void prepareBufferedClient(uint32_t buffer_size, std::chrono::milliseconds ttl) {
+    buffered_client_ =
+        std::make_unique<BufferedAsyncClient<helloworld::HelloRequest, helloworld::HelloReply>>(
+            buffer_size, *method_descriptor_, callback_, *client_, *dispatcher_, ttl);
+  }
+
+  void bufferNewMessage(absl::optional<uint32_t> expected_message_id) {
+    helloworld::HelloRequest request;
+    request.set_name("Alice");
+    EXPECT_EQ(expected_message_id, buffered_client_->bufferMessage(request));
+  }
+
+  void validateBuffer(uint32_t expected_buffered_count, uint32_t expected_pending_count) {
+    const auto buffer = buffered_client_->messageBuffer();
+    uint32_t buffered_count = 0;
+    uint32_t pending_count = 0;
+
+    for (const auto& message : buffer) {
+      switch (message.second.first) {
+      case BufferState::Buffered:
+        ++buffered_count;
+        break;
+      case BufferState::PendingFlush:
+        ++pending_count;
+        break;
+      default:
+        break;
+      }
+    }
+
+    EXPECT_EQ(buffered_count, expected_buffered_count);
+    EXPECT_EQ(pending_count, expected_pending_count);
+  }
+
+  Api::ApiPtr api_;
+  Event::DispatcherPtr dispatcher_;
   const Protobuf::MethodDescriptor* method_descriptor_;
   envoy::config::core::v3::GrpcService config_;
   NiceMock<Upstream::MockClusterManager> cm_;
   NiceMock<Http::MockAsyncClient> http_client_;
+  Http::MockAsyncClientStream http_stream_;
+  DangerousDeprecatedTestTime test_time_;
```

Comment:
Can we use simulated test time here instead?

Reply:
Simulated test time is no longer needed here. Thus I deleted it.